Silicon/ArrowlakePkg/Library/PciePm/PciePm.c

```diff
@@ -0,0 +1,348 @@
+  @param[out] RpFun                 Return corresponding root port function number.
+
+  @retval     EFI_SUCCESS           Root port device and function is retrieved
+  @retval     EFI_INVALID_PARAMETER RpNumber is invalid
+**/
+EFI_STATUS
+EFIAPI
+GetCpuPcieRpDevFun (
+  IN  UINTN   RpNumber,
+  OUT UINTN   *RpDev,
+  OUT UINTN   *RpFun
+  )
+{
+  if (RpNumber > GetMaxCpuPciePortNum ()) {
+    DEBUG ((DEBUG_ERROR, "GetCpuPcieRpDevFun invalid RpNumber %x", RpNumber));
+    ASSERT (FALSE);
+    return EFI_INVALID_PARAMETER;
+  }
+
+  switch (RpNumber) {
+    case 0:
+      *RpDev = 6;
+      *RpFun = 0;
+      break;
+    case 1:
+      *RpDev = 1;
+      *RpFun = 0;
+      break;
+    case 2:
+      if (GetCpuSku () == EnumCpuTrad) {
+        *RpDev = 1;
+        *RpFun = 1;
+      } else {
+        *RpDev = 6;
+        *RpFun = 2;
+      }
+      break;
+    default:
+      *RpDev = 6;
+      *RpFun = 0;
+      break;
+  }
+  return EFI_SUCCESS;
+}
+
+
+/**
+  Config CPU PCIE power management settings
+**/
+VOID
+CpuPciePmConfig (
+  VOID
+  )
+{
+  EFI_STATUS  Status;
+  UINT32      PortIndex;
+  UINT64      RpBase;
+  UINT8       MaxPciePortNum;
+  UINTN       RpDevice;
+  UINTN       RpFunction;
+  UINT8       TempRootPortBusNumMin;
+  UINT8       TempRootPortBusNumMax;
+
+  TempRootPortBusNumMin = 2;
+  TempRootPortBusNumMax = 10;
+
+  MaxPciePortNum = GetMaxCpuPciePortNum ();
+
+  for (PortIndex = 0; PortIndex < MaxPciePortNum; PortIndex++) {
+    Status = GetCpuPcieRpDevFun (PortIndex, &RpDevice, &RpFunction);
+    if (EFI_ERROR (Status)) {
+      continue;
+    }
+
+    RpBase = PCI_SEGMENT_LIB_ADDRESS (SA_SEG_NUM, SA_MC_BUS, (UINT32) RpDevice, (UINT32) RpFunction, 0);
+
+    if (PciSegmentRead16 (RpBase) != 0xFFFF) {
+      RootportDownstreamPmConfiguration (
+        SA_SEG_NUM,
+        SA_MC_BUS,
+        (UINT8)RpDevice,
+        (UINT8)RpFunction,
+        TempRootPortBusNumMin,
+        TempRootPortBusNumMax,
+        &mCpuPcieRootPortConfig[PortIndex].PcieRpCommonConfig,
+        0,
+        NULL
+      );
+    }
+  }
+}
+
+/**
+  Config PCH PCIE power management settings
+**/
+VOID
+PchPciePmConfig (
+  VOID
+  )
+{
+  UINT32  PortIndex;
+  UINT64  RpBase;
+  UINT8   MaxPciePortNum;
+  UINT8   TempRootPortBusNumMin;
+  UINT8   TempRootPortBusNumMax;
+
+  TempRootPortBusNumMin = 2;
+  TempRootPortBusNumMax = 10;
+
+  MaxPciePortNum = GetPchMaxPciePortNum ();
+
+  for (PortIndex = 0; PortIndex < MaxPciePortNum; PortIndex++) {
+    RpBase = PchPcieRpPciCfgBase (PortIndex);
+
+    if (PciSegmentRead16 (RpBase) != 0xFFFF) {
+      RootportDownstreamPmConfiguration (
+        DEFAULT_PCI_SEGMENT_NUMBER_PCH,
+        DEFAULT_PCI_BUS_NUMBER_PCH,
+        PchPcieRpDevNumber (PortIndex),
+        PchPcieRpFuncNumber (PortIndex),
+        TempRootPortBusNumMin,
+        TempRootPortBusNumMax,
+        &mPcieRootPortConfig[PortIndex].PcieRpCommonConfig,
+        ARRAY_SIZE (mPcieDeviceTable) - 1,
+        mPcieDeviceTable
```

Comment:
@kokweich for ARLS test, do we need to input this table?  as a placeholder, can we simple leave a null character {0} at end? thanks

Reply:
I tried leaving null character and s0ix is having some issue, hence leaving the table as it is